tactile_state_publisher/src/tactile_state_publisher/tactile_bias.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+# license removed for brevity
+import rospy
+from tactile_msgs.msg import TactileState
+from std_srvs.srv import EmptyResponse, Empty
+from copy import deepcopy
+import numpy as np
+
+SAMPLE_SIZE = 200
+
+class tactile_bias(object):
+	def __init__(self, count = SAMPLE_SIZE):
+		self.initialized = False
+		self.initial_average_count = count
+		self.average_counter = count
+		self.average_vec = []
+		self.bias = []
+		self.pub = rospy.Publisher('/tactile_states_biased', TactileState, queue_size=10)
+		self.sub = rospy.Subscriber("/tactile_states", TactileState, self.callback)
+		self.service = rospy.Service('tactile_bias/bias', Empty, self.handle_bias)
+		rospy.spin()
+
+	def handle_bias(self, req):
+		self.average_counter = self.initial_average_count
+		self.average_vec = []
+		self.initialized = False
+		return EmptyResponse()
+		
+	def compute_averages(self, vecs):
+		avgs = []
+		for vec in vecs:
+			avgs.append(np.mean(np.array(vec)))
+		return avgs
+		
+	def callback(self, data):
+		if not self.initialized:
+			if len(self.average_vec) == 0:
+				self.average_vec = [0]*len(data.sensors[0].values)
+				for i,val in enumerate(data.sensors[0].values):
+					self.average_vec[i] = []
+			# accumulate values
+			if self.average_counter > 0:
+				for i,val in enumerate(data.sensors[0].values):
+					self.average_vec[i].append(val)
+				self.average_counter -= 1
+			# enough data to compute the average
+			if self.average_counter <= 0:
+				self.bias = self.compute_averages(self.average_vec)
+				rospy.loginfo("computed bias:" + str(self.bias))
+				self.initialized = True
+		else:
+			newvals = []
+			for i,val in enumerate(data.sensors[0].values):
```

Comment:
Why don't you use efficient numpy vector calculations, i.e. `data.sensors[0] - self.bias`?
You might need to numpy-ize your messages: http://library.isr.ist.utl.pt/docs/roswiki/rospy_tutorials(2f)Tutorials(2f)numpy.html

Reply:
> You might need to numpy-ize your messages: http://library.isr.ist.utl.pt/docs/roswiki/rospy_tutorials(2f)Tutorials(2f)numpy.html

I checked, and this tutorial existed since 2009 but is not part of the beginner tutorials one does when starting ROS. I also never in 10 years of using ROS stumbled upon this type of deserialization. Either it is not well known or has a reason that it is not used.
I will definitively test various things : no numpy, on the fly numpy.asarray conversions and  numpy_msg deserialization.

Reply:
I patched the _TactileState.py message to add some timing and get it back for accumulation in my callback
```
10d9
< import datetime
70,71c69,70
<   __slots__ = ['header','sensors', 'timing']
<   _slot_types = ['std_msgs/Header','sensor_msgs/ChannelFloat32[]','datetime']
---
>   __slots__ = ['header','sensors']
>   _slot_types = ['std_msgs/Header','sensor_msgs/ChannelFloat32[]']
97d95
<     self.timing = datetime.timedelta(microseconds=0)
140d137
<     start_time = datetime.datetime.now()
183,185d179
<       end_time = datetime.datetime.now()
<       #print("std deserialize: ",end_time - start_time)
<       self.timing = end_time - start_time
228d221
<     start_time = datetime.datetime.now()
271,273d263
<       end_time = datetime.datetime.now()
<       #print("np deserialize: ", end_time - start_time)
<       self.timing = end_time - start_time
```

then timed only the computation part
```
self.start_time = datetime.datetime.now()

EITHER original code no numpy
  newvals = []
  for i,val in enumerate(data.sensors[0].values):
	newvals.append(val-self.bias[i])
OR numpy for computation but standard deserialize
  newvals = np.asarray(data.sensors[0].values) - self.bias
OR numpy for computation and numpy for deserialize using numpy_msg
  newvals = data.sensors[0].values - self.bias

self.end_time = datetime.datetime.now()
```

did 1000 measurements and averaged out both the deserialization part and the computation part

I encountered issues depending  on the rate of publication of the tactile messages, the time would differ. I don't understand why. The code is so simple I don't see where the rate comes into play. I ignored this issue and used the same rate of 1 kHz of tactile_state message for all examples

- original code
```
vec size 1: 29 us deser + 11 us compute = 40 us
vec size 2: 30 us deser + 11 us compute = 41 us
vec size 10: 30 + 15 =  45 us
vec size 20: 29 + 19 =  48 us
vec size 64: 28 + 32 = 60 us => okay does not scale up well
```
- numpy for computation but standard serialize
```
vec size 1: 29 + 24 = 53 us
vec size 2: 30 + 25 = 55 us
vec size 10: 29 + 24 = 53 us
vec size 20: 28 + 24 = 52 us
vec size 64: 27 + 25 = 52 us
```
- numpy for computation and numpy for deserialize using numpy_msg
```
vec size 1: 35 + 14 = 49 us
vec size 2: 35 + 14 = 49 us
vec size 10: 36 + 14 = 50 us
vec size 20: 34 + 14 = 48 us
vec size 64: 35 + 14 = 49 us
```

conclusion 
- numpy_msg has some benefits but not enormous if only one vector is used (takes more time to deserialize to numpy)
- numpy computation with small vector is not good, under size 20 do not bother switch to numpy_msg.

basically, my non-optimized code is 10 us slower per message of size 64
at 1 kHz it means we loose 10 ms of processing time per second.
so if running this code 200 hours, we can compensate for my 2 hours optimizing (and testing) this code.

I would say in the future one should not take for granted that numpy is faster see https://stackoverflow.com/questions/7133885/fastest-way-to-grow-a-numpy-numeric-array and always evaluate. 
But then, evaluation takes time, and this not really worth I would say. If some processing appears to consume lot of CPU or delays data or cannot keep up I would then worry and try optimize, otherwise, we are not sure unless we test.

Reply:
addressed in the latest fixes